py/desispec/io/photo.py

```diff
@@ -0,0 +1,862 @@
+                        if col in photo1.colnames:
+                            _photo[col] = photo1[col]
+                    del photo1
+                    photofiles.append('TOO')
+                    photo.append(_photo)
+                continue
+
+            # get the correct extension name or number
+            tinfo = fitsio.FITS(targetfile)
+            for _tinfo in tinfo:
+                extname = _tinfo.get_extname()
+                if 'TARGETS' in extname:
+                    break
+            if extname == '':
+                extname = 1
+                
+            # fitsio does not preserve the order of the rows but we'll sort later.
+            photo_targetid = tinfo[extname].read(columns='TARGETID')
+            I = np.where(np.isin(photo_targetid, input_cat['TARGETID']))[0]
+            
+            log.debug('Matched {} targets in {}'.format(len(I), targetfile))
+            if len(I) > 0:
+                photo1 = tinfo[extname].read(rows=I)
+                # Columns can be out of order, so sort them here based on the
+                # data model so we can stack below.
+                _photo = Table(np.hstack(np.repeat(datamodel, len(I))))
+                for col in _photo.colnames: # all these columns should exist...
+                    if col in photo1.dtype.names:
+                        _photo[col] = photo1[col]
+                    else:
+                        log.debug('Skipping missing column {} from {}'.format(col, targetfile))
+                del photo1
+                photofiles.append(targetfile)
+                photo.append(_photo)
+                #if input_cat['TILEID'][0] == 80615:
+                #    pdb.set_trace()
+
+    # backup programs have no target catalog photometry at all
+    if len(photo) == 0:
+        log.warning('No targeting photometry found.')
+        raise ValueError
+
+    # np.hstack will sometimes complain even if the tables are identical...
+    #photo = Table(np.hstack(photo))
+    photo = vstack(photo)
+
+    # make sure there are no duplicates...?
+    _, uindx = np.unique(photo['TARGETID'], return_index=True) 
+    photo = photo[uindx]
+    assert(len(np.unique(photo['TARGETID'])) == len(photo))
+
+    # sort explicitly in order to ensure order
+    I = np.where(np.isin(out['TARGETID'], photo['TARGETID']))[0]
+    srt = np.hstack([np.where(tid == photo['TARGETID'])[0] for tid in out['TARGETID'][I]])
+    out[I] = photo[srt]
+
+    #check = (out['RA'] == 0) * (out['DEC'] == 0)
+    #if np.sum(check) > 0:
+    #    log.warning('RA and DEC are missing for {} objects'.format(np.sum(check)))
+    #    #raise ValueError
+
+    if columns is not None:
+        out = out[columns]
+    
+    return out
+
+def _tractorphot_datamodel(from_file=False):
+    """Initialize the tractorphot data model for DR9 photometry."""
+
+    if from_file:
+        datamodel_file = '/global/cfs/cdirs/cosmo/data/legacysurvey/dr9/south/tractor/000/tractor-0001m002.fits'
```

Comment:
To avoid the CFS hardcode while living within the DESI envvar system, you can use `$DESI_ROOT/external/legacysurvey` which is a softlink to `/global/cfs/cdirs/cosmo/data/legacysurvey`

Reply:
Done.